Correct mixed vertical orientation of Mongolian and Phags-pa characters per UTR#50 and css-writing-modes

Ahmad-S792 opened the pull request:
#### 6c53862898c9af2cdd913b4c094fc307948ea1a6
<pre>
Correct mixed vertical orientation of Mongolian and Phags-pa characters per UTR#50 and css-writing-modes

<a href="https://bugs.webkit.org/show_bug.cgi?id=240279">https://bugs.webkit.org/show_bug.cgi?id=240279</a>
<a href="https://rdar.apple.com/problem/93426525">rdar://problem/93426525</a>

Reviewed by Alan Baradlay.

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Merge: <a href="https://chromium.googlesource.com/chromium/blink/+/e3ba3532accd84e6a0253705595b97556ae7ca67">https://chromium.googlesource.com/chromium/blink/+/e3ba3532accd84e6a0253705595b97556ae7ca67</a>

Characters for Mongolian and Phags-Pa scripts were changed to R (rotate)
from U (upright) in rev 13 of the Unicode UTR#50[1], the specification of
character orientations in vertical flow. CSS Writing Modes Level 3 refers
this specification to use in vertical flow[2].

[1] <a href="https://www.unicode.org/reports/tr50/">https://www.unicode.org/reports/tr50/</a>
[2] <a href="https://drafts.csswg.org/css-writing-modes/#vertical-orientations">https://drafts.csswg.org/css-writing-modes/#vertical-orientations</a>

* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(shouldIgnoreRotation):
* LayoutTests/TestExpectations: Remove &apos;failing&apos; expectation
* LayoutTests/platform/ios/TestExpectations: Can&apos;t add &apos;pixel&apos; tolerance so adding
&apos;ImageOnlyFailure&apos; expectation
* LayoutTests/platform/glib/TestExpectations: Changing &apos;Pass&apos; to &apos;ImageOnlyFailure&apos; expectation
due to shaping across color change element boundaries being still broken

Canonical link: <a href="https://commits.webkit.org/272454@main">https://commits.webkit.org/272454@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0aac97cb588263b92517a2e713c53d9673dfec82

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31643 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10322 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33358 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34131 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28651 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32421 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12676 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7572 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28250 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31991 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8690 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28229 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7495 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7652 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28142 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35474 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28755 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28589 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33780 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7749 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5758 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31632 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9406 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7429 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8432 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8262 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -3324,6 +3324,10 @@ imported/w3c/web-platform-tests/css/css-lists/list-style-type-string-002.html [
 # These tests are for iOS port. Should be pass.
 fast/shadow-dom/ios/ [ Pass ]
 
+# This test fails in iOS due to subpixel differences for the moongolian characters
+imported/w3c/web-platform-tests/css/css-text/i18n/ja/css-text-line-break-ja-pr-normal.html [ ImageOnlyFailure ]
+imported/w3c/web-platform-tests/css/css-text/i18n/ja/css-text-line-break-ja-pr-strict.html [ ImageOnlyFailure ]
```
if these failures here really just about subpixel diffs, it may be better to introduce fuzziness

Ahmad-S792 replied:
These are WPT tests and if I add even 'fuzziness' meta tag, on next import, it will be overridden.

alanbaradlay replied:
I am pretty sure there's a way to avoid it, we do it all the time.

Ahmad-S792 replied:
Cool! I never knew it. Good to know. Will add meta tag for pixel tolerance. 👍

alanbaradlay replied:
`grep -i -r "meta name=\"fuzzy" LayoutTests/imported/w3c/web-platform-tests/css` -> 425

alanbaradlay reviewed (approved):
let's fix fuzziness in a followup

Ahmad-S792 committed:
Correct mixed vertical orientation of Mongolian and Phags-pa characters per UTR#50 and css-writing-modes

https://bugs.webkit.org/show_bug.cgi?id=240279
rdar://problem/93426525

Reviewed by Alan Baradlay.

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Merge: https://chromium.googlesource.com/chromium/blink/+/e3ba3532accd84e6a0253705595b97556ae7ca67

Characters for Mongolian and Phags-Pa scripts were changed to R (rotate)
from U (upright) in rev 13 of the Unicode UTR#50[1], the specification of
character orientations in vertical flow. CSS Writing Modes Level 3 refers
this specification to use in vertical flow[2].

[1] https://www.unicode.org/reports/tr50/
[2] https://drafts.csswg.org/css-writing-modes/#vertical-orientations

* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(shouldIgnoreRotation):
* LayoutTests/TestExpectations: Remove 'failing' expectation
* LayoutTests/platform/ios/TestExpectations: Can't add 'pixel' tolerance so adding
'ImageOnlyFailure' expectation
* LayoutTests/platform/glib/TestExpectations: Changing 'Pass' to 'ImageOnlyFailure' expectation
due to shaping across color change element boundaries being still broken

Canonical link: https://commits.webkit.org/272454@main
